### ISSUE TITLE
util.representation now depends on blender

### ISSUE THREAD

**Issue by brunopostle** [IfcOpenShell/IfcOpenShell#4057]
With 3e268be, these two lines add a blender dependency to `util.representation` (and as a result `api.material.assign_material`, `api.type.assign_type` etc..):

    import ifcopenshell.util.shape_builder
    from mathutils import Vector

**Comment by brunopostle** [comment]
Thanks! this fix leaves a blender dependency for `geometry.add_profile_representation`, is this intended?

**Comment by Moult** [comment]
No, ideally we should use numpy.

**Comment by Moult** [comment]
A minor note that mathutils is technically Blender agnostic, but in practice it's a good idea to just use numpy for these simpler things.

**Comment by brunopostle** [comment]
Thanks again. If I install mathutils then import ifcopenshell in blender there will be two mathutils, and this is the sort of thing that causes headaches.

The new `util.data` module produces this warning, I'm not sure what this division is trying to achieve:

    /usr/lib64/python3.12/site-packages/ifcopenshell/util/data.py:87: RuntimeWarning: invalid value encountered in divide
      x_axis /= np.linalg.norm(x_axis)

**Comment by Moult** [comment]
That should normalize the x axis. Did I mess up?

**Comment by brunopostle** [comment]
It looks like you are dividing a vector by its normal which probably does weird stuff. I would expect that you would assign the normal rather than divide, or divide by the magnitude.

**Comment by Moult** [comment]
Seems alright to me https://www.stackvidhya.com/how-to-normalize-numpy-array-to-a-unit-vector/

**Comment by brunopostle** [comment]
'Frobenius Norm', I learn something every day.

I'll investigate what is causing the 'invalid value' warning.

**Comment by brunopostle** [comment]
I was fully expecting the `RuntimeWarning: invalid value encountered in divide` error to be a problem with my code, but it seems to be here: https://github.com/IfcOpenShell/IfcOpenShell/blob/c52fe8abf9cb16d6274b7a369ce6981a3a25a4bf/src/ifcopenshell-python/ifcopenshell/util/data.py#L80-L87

If I run this code with `normal = [0.0, 0.0, -1.0]` then the result is `[nan nan nan]`.

**Comment by brunopostle** [comment]
Closing as fixed in a8078e3

**Comment by aothms** [comment]
Good catch @brunopostle. This is a common thing, to pick a reference vector that is not identical to the main axis.

Instead of:

```python
if np.allclose(normal, np.array([0.0, 0.0, 1.0]), atol=1e-2) or np.allclose(
    normal, np.array([0.0, 0.0, -1.0]), atol=1e-2
):
```

We can also consider something like this:

```python
if abs(normal @ (0.,0.,1.)) > 0.9:
```

I normally use actually a rather small value such as 0.9 or even lower, because my intuition tells me it results in a bit higher fidelity, because otherwise you would end up crossing [0.03 0. 0.9995] and [0 0 1] for example, but I've no proof at all to back up whether that results into any significant loss of precision.
